source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,600 @@
+  // block).
+  uint32_t convergence_block_id = header_block->MergeBlockId();
+  while (ir_context->cfg()->preds(convergence_block_id).size() == 1) {
+    convergence_block_id = ir_context->cfg()->preds(convergence_block_id)[0];
+  }
+
+  // Get the mapping from instructions to fresh ids.
+  auto instructions_to_fresh_ids = GetInstructionsToFreshIdsMapping(ir_context);
+
+  // Keep track of the number of overflow ids used.
+  uint32_t overflow_ids_used = 0;
+
+  auto branch_instruction = header_block->terminator();
+
+  opt::BasicBlock* last_true_block = nullptr;
+
+  // Adjust the conditional branches by enclosing problematic instructions are
+  // enclosed by conditionals and get references to the last block in each
+  // branch.
+  for (int branch = 2; branch >= 1; branch--) {
+    // branch = 1 corresponds to the true branch, branch = 2 corresponds to the
+    // false branch. Consider the false branch first so that the true branch is
+    // laid out right after the false branch.
+
+    auto block = header_block;
+    // Get the id of the first block in this branch.
+    uint32_t block_id = branch_instruction->GetSingleWordInOperand(branch);
+
+    // Consider all blocks in the branch until the convergence block is reached.
+    while (block_id != convergence_block_id) {
+      // Move the block to right after the previous one.
+      block->GetParent()->MoveBasicBlockToAfter(block_id, block);
+
+      block = ir_context->cfg()->block(block_id);
+      block_id = block->terminator()->GetSingleWordInOperand(0);
+
+      // Find all the problematic instructions in the block (OpStore, OpLoad,
+      // OpFunctionCall).
+      std::vector<opt::Instruction*> problematic_instructions;
+
+      block->ForEachInst(
+          [&problematic_instructions](opt::Instruction* instruction) {
+            switch (instruction->opcode()) {
+              case SpvOpStore:
+              case SpvOpLoad:
+              case SpvOpFunctionCall:
+                problematic_instructions.push_back(instruction);
+                break;
+              default:
+                break;
+            }
+          });
+
+      uint32_t condition_id =
+          header_block->terminator()->GetSingleWordInOperand(0);
+
+      // Enclose all of the problematic instructions in conditionals, with the
+      // same condition as the selection construct being flattened.
+      for (auto instruction : problematic_instructions) {
+        // Collect the fresh ids needed by this instructions
+        uint32_t ids_needed =
+            NumOfFreshIdsNeededByOpcode(instruction->opcode());
+        std::vector<uint32_t> fresh_ids;
+
+        // Get them from the map.
+        if (instructions_to_fresh_ids.count(instruction) != 0) {
+          fresh_ids = instructions_to_fresh_ids[instruction];
+        }
+
+        // Get the ones still needed from the overflow ids.
+        for (int still_needed = ids_needed - (int)fresh_ids.size();
```

Comment:
I'm not sure about this minus operator. If I'm not mistaken, `(int)fresh_ids.size()` will be cast to `uint32_t` which might cause an overflow when subtracted.

Reply:
I think that shouldn't be a problem, since the result would then be cast to int and will thus be interpreted as signed. The subtraction operation does not operate differently on signed and unsigned integers.